website/src/_posts/2017-08-28-timely-processing.md

```diff
@@ -316,18 +343,43 @@ new DoFn<Event, EnrichedEvent>() {
   public void onExpiry(
       OnTimerContext context,
       @StateId("buffer") BagState<Event> bufferState) {
-    if (!bufferState.isEmpty().read()) {
+    if (!bufferState.read().isEmpty()) {
       for (EnrichedEvent enrichedEvent : enrichEvents(bufferState.read())) {
         context.output(enrichedEvent);
       }
+      bufferState.clear();
     }
   }
 }
 ```
 
 ```py
-# State and timers are not yet supported in Beam's Python SDK.
-# Follow https://issues.apache.org/jira/browse/BEAM-2687 for updates.
+class StatefulBufferingFn(beam.DoFn):
+  …
+
+  EXPIRY_TIMER = TimerSpec('expiry', TimeDomain.WATERMARK)
+
+  def process(self, element,
+              w=beam.DoFn.WindowParam,
+              buffer_state=beam.DoFn.StateParam(BUFFER_STATE),
+              count_state=beam.DoFn.StateParam(COUNT_STATE),
+              expiry_timer=beam.DoFn.TimerParam(EXPIRY_TIMER)):
+
+    expiry_timer.set(w.end + ALLOWED_LATENESS)
+
+    … same logic as above …
+
+  @on_timer(EXPIRY_TIMER)
+  def expiry(self,
+             buffer_state=beam.DoFn.StateParam(BUFFER_STATE),
+             count_state=beam.DoFn.StateParam(COUNT_STATE)):
+    events = buffer_state.read() or []
```

Comment:
This also strikes me as unusual. I would expect the default value to be a non-nullable (`None`-able?) iterable of some sort, just as good hygeine. So the defaulting to `[]` not necessary.

Reply:
Fixed. Thanks.